### PR TITLE
[SPARK-6876] [PySpark] [SQL] add DataFrame na.replace in pyspark

### PR DESCRIPTION
**Pull request by adrian-wang**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala [hunk 1]
```diff
@@ -53,4 +53,11 @@ private[spark] object PythonUtils {
   def toSeq[T](cols: JList[T]): Seq[T] = {
     cols.toList.toSeq
   }
+
+  /**
+   * Convert java map of K, V into Map of K, V (for calling API with varargs)
+   */
+  def toMap[K, V](jm: java.util.Map[K, V]): Map[K, V] = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala]: toScalaMap?

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -896,6 +900,77 @@ def fillna(self, value, subset=None):
 
             return DataFrame(self._jdf.na().fill(value, self._jseq(subset)), self.sql_ctx)
 
+    def replacena(self, to_replace, value, subset=None):
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: i think it's just replace in pandas.

File: core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala (hunk 1 above)
Comment (adrian-wang) [inline core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala]: yes

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (adrian-wang) [inline python/pyspark/sql/dataframe.py]: OK, I was just confused that we put `def replace()` in `DataFrameNaFunctions`.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -1182,6 +1256,13 @@ def _to_seq(sc, cols, converter=None):
     return sc._jvm.PythonUtils.toSeq(cols)
 
 
+def _to_map(sc, jm):
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: rename this _to_scala_map

File: core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala (hunk 1 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala]: sry I wasn't clear. I meant let's rename this to toScalaMap.

**Comment by rxin** [comment]
Thanks. I finally had time to review this in detail. A few things we need:

1. There are lot of branches for testing various input types. We should test those as well in tests.py. (Not in docstring tests).

2. We should support the case where to_replace is just a dict, like Pandas does.

3. Ideally, we should also support the case where to_replace is a tuple, and value is just a single value. The semantics is that for all values in to_replace, just replace it with value.

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -896,6 +900,76 @@ def fillna(self, value, subset=None):
...
+        +----+------+----+
+        | age|height|name|
+        +----+------+----+
+        |  10|    80|   A|
+        |   5|  null|   B|
+        |null|  null| Tom|
+        |null|  null|null|
+        +----+------+----+
+        """
+        if not isinstance(to_replace, (float, int, long, basestring, list, tuple)):
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: the error handling are pretty nice!

**Comment by rxin** [comment]
LGTM. Can you update it so it can merge cleanly with master?

**Comment by rxin** [comment]
Thanks. Merging in!

**Pull request closed without merging**